### PR TITLE
DaCe Libraries

### PR DESCRIPTION
**Pull request by definelicht**
We are implementing DaCe libraries to address multiple issues/lacking features:
- Allow adding custom nodes that are not "pure" DaCe, but capture important patterns
- Give users a way to build, manage, maintain, and share their DaCe extensions
- Preserve domain information in high-level nodes before "expanding" them to "pure" DaCe nodes
- Handle dependencies on external libraries, like MKL or cuBLAS

DaCe libraries will contain:
- Nodes, that each have a list on different ways to "expand" them. All library nodes must (potentially recursively) fully expanded to pure DaCe before hitting the code generator. For external function calls, such as SGEMM, we can have a shorthand that just expands into a tasklet with the code inlined.
- Transformations, responsible for expanding nodes into subgraphs in the SDFG. We can also add any other transformations here, not related to the custom nodes defined by the library.
- Environments, which define how to build and link to certain functionality (e.g., cuBLAS or FFTW), by interfacing with the DaCe CMake flow.
- Dependencies: list of other DaCe libraries that the library uses functionality from, including which version of said library.

Task list:
- [x] Allow specifying library-wide and per-node default implementation from user code (the latter overwrites the former).
- [x] Add support and test dependencies.
- [x] Support extending existing library nodes with more expansions (this almost works already, but nodes are currently registered with their expansions).
- [x] ~~Make Reduce a library node.~~
- [x] DIODE support.
- [x] Add test for frontend using matmul (w/ env vars)

### CODE REVIEW

**Comment by tbennun** [comment]
Jenkins, do your thing.

**Comment by definelicht** [comment]
@alexnick83 Perhaps we should apply strict transformations after every library node expansion automatically?

**Comment by alexnick83** [comment]
> @alexnick83 Perhaps we should apply strict transformations after every library node expansion automatically?

I added a second call to apply_strict_transformations, after all the expansions have been completed (see c829e4c).

**Comment by definelicht** [comment]
@tbennun I've done the tasks I could. Should we consider postponing pulling out reduce to a future PR?

**Comment by tbennun** [comment]
@definelicht we should probably make sure the workflow works first. You're right, we should delay reduce nodes to another PR (makes review easier)

File: dace/graph/nodes.py [hunk 1]
```diff
@@ -251,7 +270,7 @@ class Tasklet(CodeNode):
         language by the code generator.
     """
 
-    label = Property(dtype=str, desc="Name of the tasklet")
+    # label = Property(dtype=str, desc="Name of the tasklet")
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 2]
```diff
@@ -280,15 +297,15 @@ def __init__(self,
                  code_exit="",
                  location="-1",
                  debuginfo=None):
-        super(Tasklet, self).__init__(inputs or set(), outputs or set())
+        super(Tasklet, self).__init__(label, location, inputs, outputs)
 
         # Properties
-        self.label = label
+        # self.label = label
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 3]
```diff
@@ -280,15 +297,15 @@ def __init__(self,
...
         # Set the language directly
         #self.language = language
         self.code = {'code_or_block': code, 'language': language}
 
-        self.location = location
+        # self.location = location
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 4]
```diff
@@ -363,7 +380,7 @@ class NestedSDFG(CodeNode):
         @note: A nested SDFG cannot create recursion (one of its parent SDFGs).
     """
 
-    label = Property(dtype=str, desc="Name of the SDFG")
+    # label = Property(dtype=str, desc="Name of the SDFG")
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 5]
```diff
@@ -392,13 +408,13 @@ def __init__(self,
                  schedule=dtypes.ScheduleType.Default,
                  location="-1",
                  debuginfo=None):
-        super(NestedSDFG, self).__init__(inputs, outputs)
+        super(NestedSDFG, self).__init__(label, location, inputs, outputs)
 
         # Properties
-        self.label = label
+        # self.label = label
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 6]
```diff
@@ -392,13 +408,13 @@ def __init__(self,
...
         self.sdfg = sdfg
         self.schedule = schedule
-        self.location = location
+        # self.location = location
```
Comment (tbennun) [inline dace/graph/nodes.py]:
```suggestion
```

File: dace/graph/nodes.py [hunk 7]
```diff
@@ -921,3 +937,134 @@ def __label__(self, sdfg, state):
...
+
+    def draw_node(self, sdfg, state):
+        return dot.draw_node(sdfg, state, self, shape="folder")
+
+    def validate(self, sdfg, state):
+        raise ValueError("Unexpanded library node of type " + str(type(self)) +
+                         ".")
+
+
+def static_connectors(clc):
```
Comment (tbennun) [inline dace/graph/nodes.py]: This is unused. Can it be removed? @alexnick83 @definelicht

File: dace/transformation/dataflow/__init__.py [hunk 8]
```diff
@@ -33,3 +33,6 @@
 from .gpu_transform import GPUTransformMap
 from .gpu_transform_local_storage import GPUTransformLocalStorage
 from .mpi import MPITransformMap
+
+# Algorithimc
```
Comment (tbennun) [inline dace/transformation/dataflow/__init__.py]:
```suggestion
# Algorithmic
```

File: dacelets/blas/nodes/matmul.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        nested_state.add_edge(acc6, None, trans_c, '_inp', dace.Memlet.from_array('_cT', CT))
+        nested_state.add_edge(trans_c, '_out', acc3, None, dace.Memlet.from_array('_c', C))
+        nested_state.add_edge(acc4, None, tasklet, '_a', dace.Memlet.from_array('_aT', AT))
+        nested_state.add_edge(acc5, None, tasklet, '_b', dace.Memlet.from_array('_bT', BT))
+        nested_state.add_edge(tasklet, '_c', acc6, None, dace.Memlet.from_array('_cT', CT))
+        nested_node = dace.graph.nodes.NestedSDFG('_cuBLAS_MatMul_', nested_sdfg, {'_a', '_b'}, {'_c'})
+        gpu_transform_tasklet(nested_sdfg, nested_state, tasklet)
+        return nested_node
+
+    # @staticmethod
```
Comment (tbennun) [inline dacelets/blas/nodes/matmul.py]:
```suggestion
```

File: dacelets/blas/nodes/matmul.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+    # def postprocessing(sdfg, state, expansion):
```
Comment (tbennun) [inline dacelets/blas/nodes/matmul.py]:
```suggestion
```

File: dacelets/blas/nodes/matmul.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+    #     gpu_transform_tasklet(sdfg, state, expansion)
```
Comment (tbennun) [inline dacelets/blas/nodes/matmul.py]:
```suggestion
```

File: dace/graph/nodes.py (hunk 7 above)
Comment (definelicht) [reply dace/graph/nodes.py]: Sure, I like the idea, but it didn't work with our current setup. I removed it for TODO(later (maybe)).

**Pull request merged**